### PR TITLE
Feature/830 892 912 filter architecture refactor

### PR DESCRIPTION
**Pull request by thibaudteil**
* **Tickets addressed:** bsk-830-892-912
* **Review:** By commit 
* **Merge strategy:** Merge (no squash) 

## Description
Refactor the current filter architecture classes. The primary goal is to strongly type the states in order to be able to identify different components to the state vector, all the while gathering main functionalities in interfaces to minimize code duplication and facilitate testing.

The first commit creates a kalman filter interface
The second commits create or refactor the state, measurement, and dynamics classes which are implemented in set in filter implementations
The rest of the commits refactor existing filter under this interface, removing a lot of fields from the modules
Verification

## Verification
All unit tests were nearly unchanged and still pass

## Documentation
Documentation was updated and new files were added

## Future work

This approach could extend to the covariance, which is a bit more involved to type according to the state given the cross correlations

### CODE REVIEW

File: src/fswAlgorithms/_GeneralModuleFiles/kalmanFilter.cpp
Comment (patkenneally) [inline src/fswAlgorithms/_GeneralModuleFiles/kalmanFilter.cpp]: :large_orange_diamond: I think this license text for each file should be "Copyright (c) 2024, Laboratory for Atmospheric and Space Physics, University of Colorado at Boulder", but we can discuss.

File: src/fswAlgorithms/_GeneralModuleFiles/kalmanFilter.cpp
Comment (patkenneally) [inline src/fswAlgorithms/_GeneralModuleFiles/kalmanFilter.cpp]: :large_blue_circle: ... measurements and output ...

File: src/fswAlgorithms/_GeneralModuleFiles/kalmanFilter.h
Comment (patkenneally) [inline src/fswAlgorithms/_GeneralModuleFiles/kalmanFilter.h]: :large_orange_diamond: Out of date comment?

File: src/fswAlgorithms/_GeneralModuleFiles/kalmanFilter.h
Comment (patkenneally) [inline src/fswAlgorithms/_GeneralModuleFiles/kalmanFilter.h]: :large_blue_circle: could group this a few lines up with the other StateVector variables. Only if you want.

File: src/fswAlgorithms/_GeneralModuleFiles/measurementModels.h
Comment (patkenneally) [inline src/fswAlgorithms/_GeneralModuleFiles/measurementModels.h]: :large_orange_diamond: It's a clearer expression of intent to `default` these in the header, rather than the implementation file.

File: src/fswAlgorithms/_GeneralModuleFiles/dynamicModels.cpp
Comment (patkenneally) [inline src/fswAlgorithms/_GeneralModuleFiles/dynamicModels.cpp:30]: :large_orange_diamond: It's not in this commit but can line 38 in `dynamicsModels.cpp` be changed to call the static function with `DynamicsModel::rk4` rather than through the instance `this->rk4`?

File: src/fswAlgorithms/attDetermination/sunlineSRuKF/_UnitTest/test_sunlineSRuKF.py
Comment (patkenneally) [inline src/fswAlgorithms/attDetermination/sunlineSRuKF/_UnitTest/test_sunlineSRuKF.py]: delete, unused.

**Review by patkenneally — APPROVED** [review]
Legend:
:small_red_triangle_down: *Issues* to address before merge
:large_orange_diamond: *Requests* that should not block merge, but should at least be discussed
:large_blue_circle: *Recommendations* that can be ignored if desired

File: src/fswAlgorithms/_GeneralModuleFiles/kalmanFilter.cpp
Comment (thibaudteil) [reply src/fswAlgorithms/_GeneralModuleFiles/kalmanFilter.cpp]: Done

File: src/fswAlgorithms/_GeneralModuleFiles/kalmanFilter.cpp
Comment (thibaudteil) [reply src/fswAlgorithms/_GeneralModuleFiles/kalmanFilter.cpp]: fixed

File: src/fswAlgorithms/_GeneralModuleFiles/kalmanFilter.h
Comment (thibaudteil) [reply src/fswAlgorithms/_GeneralModuleFiles/kalmanFilter.h]: fixed

File: src/fswAlgorithms/_GeneralModuleFiles/kalmanFilter.h
Comment (thibaudteil) [reply src/fswAlgorithms/_GeneralModuleFiles/kalmanFilter.h]: done

File: src/fswAlgorithms/_GeneralModuleFiles/measurementModels.h
Comment (thibaudteil) [reply src/fswAlgorithms/_GeneralModuleFiles/measurementModels.h]: done everywhere

File: src/fswAlgorithms/_GeneralModuleFiles/dynamicModels.cpp
Comment (thibaudteil) [reply src/fswAlgorithms/_GeneralModuleFiles/dynamicModels.cpp:30]: I'll do it here, feels like its appropriate

File: src/fswAlgorithms/attDetermination/sunlineSRuKF/_UnitTest/test_sunlineSRuKF.py
Comment (thibaudteil) [reply src/fswAlgorithms/attDetermination/sunlineSRuKF/_UnitTest/test_sunlineSRuKF.py]: done

**Pull request merged**
